File: interview_handwrite/cpp/object_pool.cpp
// Object pool — pre-allocated buffer reuse.
//
// Whiteboard talking points:
// - Avoid repeated heap alloc/free on a hot path.
// - acquire() marks a slot in_use; release() returns it to the pool.
// - Blocking acquire spins/sleeps when pool is exhausted.
// - Identity check on release prevents double-free style bugs.

#include <cassert>
#include <chrono>
#include <functional>
#include <iostream>
#include <mutex>
#include <optional>
#include <stdexcept>
#include <thread>
#include <vector>

template <typename T>
class ObjectPool {
 public:
    ObjectPool(std::function<T()> factory, size_t size) : items_(size) {
        for (size_t i = 0; i < size; ++i) {
            items_[i].value = factory();
            items_[i].in_use = false;
        }
    }

    T* acquire(bool block = true,
               std::chrono::milliseconds timeout = std::chrono::milliseconds::max()) {
        const auto deadline = std::chrono::steady_clock::now() + timeout;
        while (true) {
            {
                std::lock_guard<std::mutex> lock(mutex_);
                for (auto& item : items_) {
                    if (!item.in_use) {
                        item.in_use = true;
                        return &item.value;
                    }
                }
            }

            if (!block) {
                return nullptr;
            }

            if (timeout != std::chrono::milliseconds::max()) {
                const auto remaining = deadline - std::chrono::steady_clock::now();
                if (remaining <= std::chrono::steady_clock::duration::zero()) {
                    return nullptr;
                }
                const auto sleep_for =
                    remaining < std::chrono::milliseconds(1) ? remaining
                                                           : std::chrono::milliseconds(1);
                std::this_thread::sleep_for(sleep_for);
            } else {
                std::this_thread::sleep_for(std::chrono::milliseconds(1));
            }
        }
    }

    void release(T* obj) {
        std::lock_guard<std::mutex> lock(mutex_);
        for (auto& item : items_) {
            if (&item.value == obj) {
                item.in_use = false;
                return;
            }
        }
        throw std::invalid_argument("object does not belong to this pool");
    }

 private:
    struct PooledObject {
        T value{};
        bool in_use = false;
    };

    std::vector<PooledObject> items_;
    std::mutex mutex_;
};

bool test_object_pool() {
    ObjectPool<std::vector<char>> pool(
        [] { return std::vector<char>(1024); }, 2);

    auto* b1 = pool.acquire();
    auto* b2 = pool.acquire();
    assert(pool.acquire(false) == nullptr);

    pool.release(b1);
    auto* b3 = pool.acquire();
    assert(b3 == b1);

    pool.release(b2);
    pool.release(b3);
    return true;
}

int main() {
    assert(test_object_pool());
    std::cout << "object_pool: ok\n";
    return 0;
}

File: concurrency/object_pool.py
"""Object pool — pre-allocated buffer reuse."""

from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Generic, List, Optional, TypeVar

import threading

T = TypeVar("T")


@dataclass
class PooledObject(Generic[T]):
    value: T
    in_use: bool = False


class ObjectPool(Generic[T]):
    def __init__(self, factory, size: int):
        self._items: List[PooledObject[T]] = [PooledObject(factory()) for _ in range(size)]
        self._lock = threading.Lock()

    def acquire(self, block: bool = True, timeout: Optional[float] = None) -> Optional[T]:
        deadline = None if timeout is None else time.monotonic() + timeout
        while True:
            with self._lock:
                for item in self._items:
                    if not item.in_use:
                        item.in_use = True
                        return item.value
            if not block:
                return None
            if timeout is not None:
                remaining = deadline - time.monotonic()
                if remaining <= 0:
                    return None
                time.sleep(min(0.001, remaining))
            else:
                time.sleep(0.001)

    def release(self, obj: T) -> None:
        with self._lock:
            for item in self._items:
                if item.value is obj:
                    item.in_use = False
                    return
            raise ValueError("object does not belong to this pool")


def _test() -> None:
    pool = ObjectPool(lambda: bytearray(1024), size=2)
    b1 = pool.acquire()
    b2 = pool.acquire()
    assert pool.acquire(block=False) is None
    pool.release(b1)
    b3 = pool.acquire()
    assert b3 is b1
    pool.release(b2)
    pool.release(b3)
    print("object_pool: ok")


if __name__ == "__main__":
    _test()
